### PR TITLE
BeagleBone Firmware upgrade to Debian 11.7 Bullseye

### PR DESCRIPTION
**Pull request by Ga-Ol-St**
1. Updated PRU libraries
2. Updated PRU code
3. Resolved issues with Linker
4. Resolved issues with Kernel 5.10 requirements
5. Resolved issue with klipper_pru startup before PRU is ready
6. Resolved issue with conflicts with AVR packages (did workaround)
7. Updated Documentation for BeagleBone:
   - Added additional details in deployment steps.
   - Added multiple use cases for BeagleBone deployment.
   - Added bus configurations CAN, UART, I2C, SPI

**Impact: PRU firmware require Kernel 5.10+, old kernels are not compatible anymore.**

Related Topics in forum:
https://klipper.discourse.group/t/beaglebone-on-debian-11-7-searching-testers/14076
https://klipper.discourse.group/t/beaglebone-black-spi-gy-291-adxl345/15411

Signed-off-by: Oleg Gavavka <work@datalink.net.ua>

### CODE REVIEW

File: src/pru/Makefile [hunk 1]
```diff
@@ -27,7 +27,11 @@ target-y += $(OUT)pru0.elf $(OUT)pru1.elf
 $(OUT)pru0.elf: $(patsubst %.c, $(OUT)src/%.o,$(pru0-y))
 	@echo "  Linking $@"
 	$(Q)$(CC) $(CFLAGS_klipper.elf) $^ -o $(OUT)pru0.o
+	# dumping info about INTC
+	$(Q)$(OBJCOPY) --dump-section '.pru_irq_map'=$(OUT)pru0.pru_irq_map.bin $(OUT)src/pru/pru0.o $(OUT)pru0.o.discard
 	$(Q)$(CC) $(CFLAGS_pru0.elf) $(OUT)pru0.o -o $@
+	# in previous step linker did loose info about INTC during optimization, restoring it here (P.S. pru.lds is not helping)
+	$(Q)$(OBJCOPY) --add-section '.pru_irq_map'=$(OUT)pru0.pru_irq_map.bin $@
```
Comment (KevinOConnor) [inline src/pru/Makefile:34]: Can you describe what this does?  Could modifying the `src/pru/pru.lds` linker script accomplish the same goal instead?

File: src/pru/main.c [hunk 2]
```diff
@@ -7,7 +7,8 @@
 #include <stdint.h> // uint32_t
 #include <pru/io.h> // read_r31
 #include <pru_iep.h> // CT_IEP
-#include <pru_intc.h> // CT_INTC
+
+//#include <pru_intc.h> // PRU_INTC
```
Comment (KevinOConnor) [inline src/pru/main.c:11]: Looks like dead code.

File: src/pru/main.c [hunk 3]
```diff
@@ -75,25 +76,26 @@ timer_kick(void)
...
 static uint32_t in_timer_dispatch;
 
 static void
 _irq_poll(void)
 {
-    uint32_t secr0 = CT_INTC.SECR0;
+    uint32_t secr0 = PRU_INTC.SECR0;
     if (secr0 & (1 << KICK_PRU1_EVENT)) {
-        CT_INTC.SECR0 = 1 << KICK_PRU1_EVENT;
+    //bit 19 was active
```
Comment (KevinOConnor) [inline src/pru/main.c:89]: I'm not sure what this comment means - is it dead code?

**Review by KevinOConnor — COMMENTED** [review]
Thanks.  In general it seems fine.  I have a few comments/questions - see below.

-Kevin

File: lib/pru_rpmsg/include/pru_rpmsg.h
[the remark this answers is not in the document]
Comment (Ga-Ol-St) [reply lib/pru_rpmsg/include/pru_rpmsg.h:1]: it's from same sources, just fresh versions.
I will recheck if there is some branch reference.

File: scripts/install-debian.sh [hunk 4]
```diff
@@ -17,7 +17,8 @@ install_packages()
     # hub-ctrl
     PKGLIST="${PKGLIST} libusb-dev"
     # AVR chip installation and building
-    PKGLIST="${PKGLIST} avrdude gcc-avr binutils-avr avr-libc"
+# AVR package conflicts with gcc package, disabling dependencies for now
+#    PKGLIST="${PKGLIST} avrdude gcc-avr binutils-avr avr-libc"
```
[the remark this answers is not in the document]
Comment (Ga-Ol-St) [reply scripts/install-debian.sh:21]: Good catch, yes it should be installing AVR packages as install-beaglebone.sh will remove them

File: src/pru/Makefile (hunk 1 above)
Comment (Ga-Ol-St) [reply src/pru/Makefile:34]: Believe me I did try everything what i know  to fix this issue by changing pru.lds - but i was still losing ".pru_irq_map" - so i made this workaround.

File: src/pru/main.c (hunk 2 above)
Comment (Ga-Ol-St) [reply src/pru/main.c:11]: yepp, will fix it.

File: lib/pru_rpmsg/include/pru_rpmsg.h
[the remark this answers is not in the document]
Comment (Ga-Ol-St) [reply lib/pru_rpmsg/include/pru_rpmsg.h:1]: Fixed, updated revision number.

File: scripts/install-debian.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ga-Ol-St) [reply scripts/install-debian.sh]: Restored to original state.

File: src/pru/intc_map_0.h
[the remark this answers is not in the document]
Comment (Ga-Ol-St) [reply src/pru/intc_map_0.h:1]: This file contain PRU_IRQ_MAP definition which is required for RemoteProc in Kernel 5.10+ to properly setup interrupts for EIP and RPMsg.
Previously this information was in "pru_intc_map" variable which was included in Resource table, new kernel require separate ".pru_irq_map" resource for this.

File: src/pru/main.c (hunk 2 above)
Comment (Ga-Ol-St) [reply src/pru/main.c]: Fixed.

File: src/pru/main.c (hunk 3 above)
Comment (Ga-Ol-St) [reply src/pru/main.c]: Removed.

File: docs/Beaglebone.md [hunk 5]
```diff
@@ -6,23 +6,64 @@ PRU.
...
+Use case 1: Use BeagleBone only as a host system to run Klipper and additional software
+like Oktoprint/Fluidd + Moonraker/...  and this configuration will be driving
+external micro-controllers via serial/usb/canbus connections.
+
+Use case 2: Use BeagleBone with extension board (cape) like CRAMPS board.
+in this configuration BeagleBone will host Klipper + additional software, and
+it will drive extension board with BeagleBone PRU cores (2 additional cores 200Mh, 32Bit).
+
+Use case 3: It's same as "Use case 1" but additionally you want to drive
+BeagleBone GPIOs with high speed by utilizing PRU cores to offload main CPU.
```
[the remark this answers is not in the document]
Comment (Ga-Ol-St) [reply docs/Beaglebone.md]: Fixed.

**Comment by KevinOConnor** [comment]
Thanks.

-Kevin

**Comment by gaby64** [comment]
Im having issues getting klipper setup with the beaglebone black and replicape, can anyone else verify this works with replicape?

**Comment by gaby64** [comment]
Does anyone else have a Replicape that could test if everything still works fine, im facing issues.

https://klipper.discourse.group/t/return-of-replicape-motor-disable-issue/16637

**Pull request merged**